{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "provenance": []
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "code",
      "execution_count": 2,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "NhEzWDzAqS3F",
        "outputId": "20ce187b-4d37-45c1-d1d0-030b1c353b95"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "For example, a normalization Protocol may involve a ‘ robust multiarray average (RMA) normalization ’ Action that uses Bioconductor ‘ affy ’ Software [43] with certain Para- meterValues . Data types Description , BibliographicReferences , DatabaseEntry , URI , and FileAttachment enable researchers to freely add additional annotations to certain data types - DimensionElement , Investigation , Protocol , ProtocolApplication , and Data . For example, researchers can annotate a Gene with one or more DatabaseEntries , referring to unique database accession numbers for automated data integration. A unique feature of XGAP is the uniform treatment of the various trait and subject annotations. The drawback of allowing users to freely add additional annotations such as described above is that users and tools using metabolite and gene traits, for example, would have to inspect each Trait instance to see whether it is actually a metabolite or gene, and how it is annotated. That is why we instead use the object-oriented method of ‘ inheritance ’ to explicitly add essential properties to Trait and Subject variants to make sure that they are described in a uniform way. For example, Metabolite extends Trait , which explicitly adds properties ID, Name and Type (inherited from DimensionElement ) to metabolite specific properties Mass, Formula and Structure. See Jones et al. [38] for the complete FuGE specifi- cations and Jones and Paton [44] for a discussion on the benefits and drawbacks of alternative mechanisms for supporting extension in object models. Table 4 illustrates the usage of these annotation data types. Another feature of XGAP is the uniform treatment of all data on these subjects and traits. To understand basic data in XGAP, newcomers just have to learn that all data are stored as Data matrices with each DataElement describing an observation on Subjects and/or Traits (rows × columns). Unlike the proven matrix structures used in MAGE-TAB (tabular format for microarray gene expression experiments) [45], in XGAP these data can be on any Trait and/or Subject combination, that is, we did not create many variants of DataElement to accommodate each combination of Trait and Subject such as MAGE-TAB ’ s ExpressionDataElement (Probe × Sample), MassSpecDataElement (MassPeak × Sample), eQtlMappingDataElement (Marker × Probe), and so on. Instead, we store all these data using the generic type DataElement and limit extension to Trait and Subject only. This avoids the (combinatorial) explo- sion of DataElement extensions so researchers can provide basic data as common data matrices (of DataElements ) and can still add particular annotations flexibly to the matrix row and columns to allow for (new) biotechnologies as demonstrated in the various Trait extensions in Figure 1. Keeping this simple and uniform data structure greatly enhances data and software (re)usability and hence productivity, in line with the findings by Brazma et al. [29] and Rayner et al. [45] that the simple tabular structures underlying biological data should be exploited instead of making it overly complicated. After structural homogenization, such as provided by FuGE and XGAP, semantic queries are the remaining major barrier for integration of experimental metadata. This requires ontologies that describe the properties of the materials and also descriptions of experimental pro- cesses, data and instruments. The former are provided by species-specific ontologies that are available from various sources. The Ontology for BioMedical investigation [46] may provide a solution for the experimental descriptors and is being used in this context by, for example, the Immune Epitope Database [47]. To enable researchers to use these well understood descriptors, XGAP inherits from FuGE the mechanism of ‘ annotations ’ , a special field to link any data object to one or more ontology terms. For example, researchers can annotate a Gene with one or more OntologyTerms if required, referring to standard ontology terms from OBO [39] or ontology terms defined locally. To enable data exchange using the XGAP model, we produced a simple text-file format (XGAP-TAB) based on the experience that for data formats to be used, data files should be easily created using simple Excel and text editor tools and closely resemble existing practices. This format is automatically derived from the model by requiring that all annotations on Investigations , Protocols , Traits , Subjects , and extensions thereof, are described as delimited text files (one file per data type) with columns matching the properties described in the object model and each row describing one data instance. Optionally, sets of DataElements can also be formatted as separate text matrices with row and column names matching these in the Trait and Subject annotation files, and with each matrix value matching one DataElement . The dimensions of each data matrix are then listed by a row in the annotations on Data . Figure 2 shows one investigation in the XGAP tabular data format with one delimited text file per data type that is, there are files named ‘ probe.txt ’ and ‘ individual. txt ’ , with each row describing a microarray probe or individual, respectively - and one text matrix file per set of DataElements - that is, there are files named ‘ data/ expressions.txt ’ and ‘ data/genotypes.txt ’ . The properties of each data matrix is then described in ‘ data.txt ’ ; that is, for the ‘ data/expressions.txt ’ there is a row in ‘ data. txt ’ that says that its columns refer to ‘ individual.txt ’ , that its rows refer to ‘ probe.txt ’ and that its values are ‘ decimal ’ . Raw data sets and data sets in other formats can be retained in a directory labeled ‘ original ’ . After proving its value in several proprietary projects, a growing array of public data sets are now available at [48] demonstrating the use of XGAP-TAB [8,11,13,14,49,50]. A pilot software infrastructure is available at [51] to help genotype-to-phenotype researchers to adopt XGAP as a backbone for their data and tool integration. We chose to use the MOLGENIS toolkit (biosoftware generator for MOLecular GENetics Information Systems; see Materials and methods) to auto-generate from the XGAP model: 1, an SQL (Structured Query Language for relational databases) file with all necessary statements for setting up your own, customized variant of the XGAP database; 2, application programming interfaces (APIs) in R, Java and Web Services that allow bioinformaticians to plug-in their R processing scripts, Taverna workflows [25,52,53] and other tools; 3, a bespoke web-based graphical user interface (GUI) by which researchers can submit and retrieve data and run plugged-in tools; and 4, import/export wizards to (un) load and validate data sets exchanged in XGAP-TAB format. The auto-generation process can be repeated to quickly customize XGAP from an extended model, for example, to accommodate a particular new type of measurement technology or experimental design. Figure 3 shows the GUI to upload, manage, find and download genotype and phenotype data to the database. The GUI is generated with a uniform ‘ look-and-feel ’ , thereby lowering the barrier for novice users. Investigations can be described with all subjects, traits, data and protocol applications involved (1). (The numbers refer to steps in the figure.) Data can be entered using either the edit boxes or using menu-option ‘ file|upload ’ (2). This option enables upload of whole lists of traits and subjects from a simple tab-delimited format (3), which can easily be produced with Excel or R; MOLGENIS automatically generates online documentation describing the expected format (4). Subsequently, the protocol applications involved can be added with the resulting raw data (for example, genetic fingerprints, expression profiles) and processed data (for example, normalized profiles, QTL profiles, metabolic networks). These data can be uploaded, again using the common tab-delimited format or custom parsers (5) that bioinformaticians can ‘ plug-in ’ for specific file formats (for example, Affymetrix CEL files). The software behind the GUI checks the relationships between subjects, traits, and data elements so no ‘ orphaned ’ data are loaded into the database - for example, genetic fingerprint data cannot be added before all information is uploaded on the markers and subjects involved. Standard paths through the data upload process are employed to ensure that only complete and valid data are uploaded and to provide a consistent user experience. Biologists can use the graphical user interface to navi- gate and retrieve available data for analysis. They can use the advanced search options (6) to find certain traits, subjects, or data. Using menu option ‘ file|down- load ’ (7) they can download visible/selected (8) data as tab-delimited files to analyze them in third party software. Bioinformaticians can ‘ plug-in ’ a custom-built screen (see ‘ customization ’ section) that allows processing of selected data inside the GUI, for example, visua- lizing a correlation matrix as a graph (9) without the additional steps of downloading data and uploading it into another tool. Biologists can create link-outs to related information, for example, to probes in GeneNet- work.org (not shown). Table 5 summarizes use cases of the graphical user interface. De facto standard analysis tools are emerging, for example, tools for transcript data [20,21,24] or metabolite abundance data [22] to mention just a few. These tools are typically implemented using the open source software for statistical analysis and graphics named R [19]. Bioinformaticians can connect their particular R or Java programs to the XGAP database using an API with similar functionality to the GUI, that is, using simple commands like ‘ find ’ , ‘ add ’ and ‘ update ’ (R/API, Java/ API). Scripts in other programming languages and workflow tools like Taverna [53] can use \n"
          ]
        }
      ],
      "source": [
        "with open(\"/ganesh.txt\",'r') as file:\n",
        "  data = file.read()\n",
        "  print(data)\n"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "with open(\"/ganesh.txt\",'r') as file:\n",
        "  data = file.read()\n",
        "  split_data = data.split(\"‘ ’\")\n",
        "  print(split_data)"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "MMpNykNgqw1Z",
        "outputId": "39adfb19-9ae6-4b18-f002-c08b8cae3889"
      },
      "execution_count": 20,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "['For example, a normalization Protocol may involve a ‘ robust multiarray average (RMA) normalization ’ Action that uses Bioconductor ‘ affy ’ Software [43] with certain Para- meterValues . Data types Description , BibliographicReferences , DatabaseEntry , URI , and FileAttachment enable researchers to freely add additional annotations to certain data types - DimensionElement , Investigation , Protocol , ProtocolApplication , and Data . For example, researchers can annotate a Gene with one or more DatabaseEntries , referring to unique database accession numbers for automated data integration. A unique feature of XGAP is the uniform treatment of the various trait and subject annotations. The drawback of allowing users to freely add additional annotations such as described above is that users and tools using metabolite and gene traits, for example, would have to inspect each Trait instance to see whether it is actually a metabolite or gene, and how it is annotated. That is why we instead use the object-oriented method of ‘ inheritance ’ to explicitly add essential properties to Trait and Subject variants to make sure that they are described in a uniform way. For example, Metabolite extends Trait , which explicitly adds properties ID, Name and Type (inherited from DimensionElement ) to metabolite specific properties Mass, Formula and Structure. See Jones et al. [38] for the complete FuGE specifi- cations and Jones and Paton [44] for a discussion on the benefits and drawbacks of alternative mechanisms for supporting extension in object models. Table 4 illustrates the usage of these annotation data types. Another feature of XGAP is the uniform treatment of all data on these subjects and traits. To understand basic data in XGAP, newcomers just have to learn that all data are stored as Data matrices with each DataElement describing an observation on Subjects and/or Traits (rows × columns). Unlike the proven matrix structures used in MAGE-TAB (tabular format for microarray gene expression experiments) [45], in XGAP these data can be on any Trait and/or Subject combination, that is, we did not create many variants of DataElement to accommodate each combination of Trait and Subject such as MAGE-TAB ’ s ExpressionDataElement (Probe × Sample), MassSpecDataElement (MassPeak × Sample), eQtlMappingDataElement (Marker × Probe), and so on. Instead, we store all these data using the generic type DataElement and limit extension to Trait and Subject only. This avoids the (combinatorial) explo- sion of DataElement extensions so researchers can provide basic data as common data matrices (of DataElements ) and can still add particular annotations flexibly to the matrix row and columns to allow for (new) biotechnologies as demonstrated in the various Trait extensions in Figure 1. Keeping this simple and uniform data structure greatly enhances data and software (re)usability and hence productivity, in line with the findings by Brazma et al. [29] and Rayner et al. [45] that the simple tabular structures underlying biological data should be exploited instead of making it overly complicated. After structural homogenization, such as provided by FuGE and XGAP, semantic queries are the remaining major barrier for integration of experimental metadata. This requires ontologies that describe the properties of the materials and also descriptions of experimental pro- cesses, data and instruments. The former are provided by species-specific ontologies that are available from various sources. The Ontology for BioMedical investigation [46] may provide a solution for the experimental descriptors and is being used in this context by, for example, the Immune Epitope Database [47]. To enable researchers to use these well understood descriptors, XGAP inherits from FuGE the mechanism of ‘ annotations ’ , a special field to link any data object to one or more ontology terms. For example, researchers can annotate a Gene with one or more OntologyTerms if required, referring to standard ontology terms from OBO [39] or ontology terms defined locally. To enable data exchange using the XGAP model, we produced a simple text-file format (XGAP-TAB) based on the experience that for data formats to be used, data files should be easily created using simple Excel and text editor tools and closely resemble existing practices. This format is automatically derived from the model by requiring that all annotations on Investigations , Protocols , Traits , Subjects , and extensions thereof, are described as delimited text files (one file per data type) with columns matching the properties described in the object model and each row describing one data instance. Optionally, sets of DataElements can also be formatted as separate text matrices with row and column names matching these in the Trait and Subject annotation files, and with each matrix value matching one DataElement . The dimensions of each data matrix are then listed by a row in the annotations on Data . Figure 2 shows one investigation in the XGAP tabular data format with one delimited text file per data type that is, there are files named ‘ probe.txt ’ and ‘ individual. txt ’ , with each row describing a microarray probe or individual, respectively - and one text matrix file per set of DataElements - that is, there are files named ‘ data/ expressions.txt ’ and ‘ data/genotypes.txt ’ . The properties of each data matrix is then described in ‘ data.txt ’ ; that is, for the ‘ data/expressions.txt ’ there is a row in ‘ data. txt ’ that says that its columns refer to ‘ individual.txt ’ , that its rows refer to ‘ probe.txt ’ and that its values are ‘ decimal ’ . Raw data sets and data sets in other formats can be retained in a directory labeled ‘ original ’ . After proving its value in several proprietary projects, a growing array of public data sets are now available at [48] demonstrating the use of XGAP-TAB [8,11,13,14,49,50]. A pilot software infrastructure is available at [51] to help genotype-to-phenotype researchers to adopt XGAP as a backbone for their data and tool integration. We chose to use the MOLGENIS toolkit (biosoftware generator for MOLecular GENetics Information Systems; see Materials and methods) to auto-generate from the XGAP model: 1, an SQL (Structured Query Language for relational databases) file with all necessary statements for setting up your own, customized variant of the XGAP database; 2, application programming interfaces (APIs) in R, Java and Web Services that allow bioinformaticians to plug-in their R processing scripts, Taverna workflows [25,52,53] and other tools; 3, a bespoke web-based graphical user interface (GUI) by which researchers can submit and retrieve data and run plugged-in tools; and 4, import/export wizards to (un) load and validate data sets exchanged in XGAP-TAB format. The auto-generation process can be repeated to quickly customize XGAP from an extended model, for example, to accommodate a particular new type of measurement technology or experimental design. Figure 3 shows the GUI to upload, manage, find and download genotype and phenotype data to the database. The GUI is generated with a uniform ‘ look-and-feel ’ , thereby lowering the barrier for novice users. Investigations can be described with all subjects, traits, data and protocol applications involved (1). (The numbers refer to steps in the figure.) Data can be entered using either the edit boxes or using menu-option ‘ file|upload ’ (2). This option enables upload of whole lists of traits and subjects from a simple tab-delimited format (3), which can easily be produced with Excel or R; MOLGENIS automatically generates online documentation describing the expected format (4). Subsequently, the protocol applications involved can be added with the resulting raw data (for example, genetic fingerprints, expression profiles) and processed data (for example, normalized profiles, QTL profiles, metabolic networks). These data can be uploaded, again using the common tab-delimited format or custom parsers (5) that bioinformaticians can ‘ plug-in ’ for specific file formats (for example, Affymetrix CEL files). The software behind the GUI checks the relationships between subjects, traits, and data elements so no ‘ orphaned ’ data are loaded into the database - for example, genetic fingerprint data cannot be added before all information is uploaded on the markers and subjects involved. Standard paths through the data upload process are employed to ensure that only complete and valid data are uploaded and to provide a consistent user experience. Biologists can use the graphical user interface to navi- gate and retrieve available data for analysis. They can use the advanced search options (6) to find certain traits, subjects, or data. Using menu option ‘ file|down- load ’ (7) they can download visible/selected (8) data as tab-delimited files to analyze them in third party software. Bioinformaticians can ‘ plug-in ’ a custom-built screen (see ‘ customization ’ section) that allows processing of selected data inside the GUI, for example, visua- lizing a correlation matrix as a graph (9) without the additional steps of downloading data and uploading it into another tool. Biologists can create link-outs to related information, for example, to probes in GeneNet- work.org (not shown). Table 5 summarizes use cases of the graphical user interface. De facto standard analysis tools are emerging, for example, tools for transcript data [20,21,24] or metabolite abundance data [22] to mention just a few. These tools are typically implemented using the open source software for statistical analysis and graphics named R [19]. Bioinformaticians can connect their particular R or Java programs to the XGAP database using an API with similar functionality to the GUI, that is, using simple commands like ‘ find ’ , ‘ add ’ and ‘ update ’ (R/API, Java/ API). Scripts in other programming languages and workflow tools like Taverna [53] can use ']\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "with open(\"/ganesh.txt\",'r') as file:\n",
        "  for line in file:\n",
        "    if 'unwanted_data' not in line:\n",
        "      print(line)"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "yVL8arxxuc_U",
        "outputId": "cd7962d8-2339-480c-95a7-12eaa71c5512"
      },
      "execution_count": 3,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "For example, a normalization Protocol may involve a ‘ robust multiarray average (RMA) normalization ’ Action that uses Bioconductor ‘ affy ’ Software [43] with certain Para- meterValues . Data types Description , BibliographicReferences , DatabaseEntry , URI , and FileAttachment enable researchers to freely add additional annotations to certain data types - DimensionElement , Investigation , Protocol , ProtocolApplication , and Data . For example, researchers can annotate a Gene with one or more DatabaseEntries , referring to unique database accession numbers for automated data integration. A unique feature of XGAP is the uniform treatment of the various trait and subject annotations. The drawback of allowing users to freely add additional annotations such as described above is that users and tools using metabolite and gene traits, for example, would have to inspect each Trait instance to see whether it is actually a metabolite or gene, and how it is annotated. That is why we instead use the object-oriented method of ‘ inheritance ’ to explicitly add essential properties to Trait and Subject variants to make sure that they are described in a uniform way. For example, Metabolite extends Trait , which explicitly adds properties ID, Name and Type (inherited from DimensionElement ) to metabolite specific properties Mass, Formula and Structure. See Jones et al. [38] for the complete FuGE specifi- cations and Jones and Paton [44] for a discussion on the benefits and drawbacks of alternative mechanisms for supporting extension in object models. Table 4 illustrates the usage of these annotation data types. Another feature of XGAP is the uniform treatment of all data on these subjects and traits. To understand basic data in XGAP, newcomers just have to learn that all data are stored as Data matrices with each DataElement describing an observation on Subjects and/or Traits (rows × columns). Unlike the proven matrix structures used in MAGE-TAB (tabular format for microarray gene expression experiments) [45], in XGAP these data can be on any Trait and/or Subject combination, that is, we did not create many variants of DataElement to accommodate each combination of Trait and Subject such as MAGE-TAB ’ s ExpressionDataElement (Probe × Sample), MassSpecDataElement (MassPeak × Sample), eQtlMappingDataElement (Marker × Probe), and so on. Instead, we store all these data using the generic type DataElement and limit extension to Trait and Subject only. This avoids the (combinatorial) explo- sion of DataElement extensions so researchers can provide basic data as common data matrices (of DataElements ) and can still add particular annotations flexibly to the matrix row and columns to allow for (new) biotechnologies as demonstrated in the various Trait extensions in Figure 1. Keeping this simple and uniform data structure greatly enhances data and software (re)usability and hence productivity, in line with the findings by Brazma et al. [29] and Rayner et al. [45] that the simple tabular structures underlying biological data should be exploited instead of making it overly complicated. After structural homogenization, such as provided by FuGE and XGAP, semantic queries are the remaining major barrier for integration of experimental metadata. This requires ontologies that describe the properties of the materials and also descriptions of experimental pro- cesses, data and instruments. The former are provided by species-specific ontologies that are available from various sources. The Ontology for BioMedical investigation [46] may provide a solution for the experimental descriptors and is being used in this context by, for example, the Immune Epitope Database [47]. To enable researchers to use these well understood descriptors, XGAP inherits from FuGE the mechanism of ‘ annotations ’ , a special field to link any data object to one or more ontology terms. For example, researchers can annotate a Gene with one or more OntologyTerms if required, referring to standard ontology terms from OBO [39] or ontology terms defined locally. To enable data exchange using the XGAP model, we produced a simple text-file format (XGAP-TAB) based on the experience that for data formats to be used, data files should be easily created using simple Excel and text editor tools and closely resemble existing practices. This format is automatically derived from the model by requiring that all annotations on Investigations , Protocols , Traits , Subjects , and extensions thereof, are described as delimited text files (one file per data type) with columns matching the properties described in the object model and each row describing one data instance. Optionally, sets of DataElements can also be formatted as separate text matrices with row and column names matching these in the Trait and Subject annotation files, and with each matrix value matching one DataElement . The dimensions of each data matrix are then listed by a row in the annotations on Data . Figure 2 shows one investigation in the XGAP tabular data format with one delimited text file per data type that is, there are files named ‘ probe.txt ’ and ‘ individual. txt ’ , with each row describing a microarray probe or individual, respectively - and one text matrix file per set of DataElements - that is, there are files named ‘ data/ expressions.txt ’ and ‘ data/genotypes.txt ’ . The properties of each data matrix is then described in ‘ data.txt ’ ; that is, for the ‘ data/expressions.txt ’ there is a row in ‘ data. txt ’ that says that its columns refer to ‘ individual.txt ’ , that its rows refer to ‘ probe.txt ’ and that its values are ‘ decimal ’ . Raw data sets and data sets in other formats can be retained in a directory labeled ‘ original ’ . After proving its value in several proprietary projects, a growing array of public data sets are now available at [48] demonstrating the use of XGAP-TAB [8,11,13,14,49,50]. A pilot software infrastructure is available at [51] to help genotype-to-phenotype researchers to adopt XGAP as a backbone for their data and tool integration. We chose to use the MOLGENIS toolkit (biosoftware generator for MOLecular GENetics Information Systems; see Materials and methods) to auto-generate from the XGAP model: 1, an SQL (Structured Query Language for relational databases) file with all necessary statements for setting up your own, customized variant of the XGAP database; 2, application programming interfaces (APIs) in R, Java and Web Services that allow bioinformaticians to plug-in their R processing scripts, Taverna workflows [25,52,53] and other tools; 3, a bespoke web-based graphical user interface (GUI) by which researchers can submit and retrieve data and run plugged-in tools; and 4, import/export wizards to (un) load and validate data sets exchanged in XGAP-TAB format. The auto-generation process can be repeated to quickly customize XGAP from an extended model, for example, to accommodate a particular new type of measurement technology or experimental design. Figure 3 shows the GUI to upload, manage, find and download genotype and phenotype data to the database. The GUI is generated with a uniform ‘ look-and-feel ’ , thereby lowering the barrier for novice users. Investigations can be described with all subjects, traits, data and protocol applications involved (1). (The numbers refer to steps in the figure.) Data can be entered using either the edit boxes or using menu-option ‘ file|upload ’ (2). This option enables upload of whole lists of traits and subjects from a simple tab-delimited format (3), which can easily be produced with Excel or R; MOLGENIS automatically generates online documentation describing the expected format (4). Subsequently, the protocol applications involved can be added with the resulting raw data (for example, genetic fingerprints, expression profiles) and processed data (for example, normalized profiles, QTL profiles, metabolic networks). These data can be uploaded, again using the common tab-delimited format or custom parsers (5) that bioinformaticians can ‘ plug-in ’ for specific file formats (for example, Affymetrix CEL files). The software behind the GUI checks the relationships between subjects, traits, and data elements so no ‘ orphaned ’ data are loaded into the database - for example, genetic fingerprint data cannot be added before all information is uploaded on the markers and subjects involved. Standard paths through the data upload process are employed to ensure that only complete and valid data are uploaded and to provide a consistent user experience. Biologists can use the graphical user interface to navi- gate and retrieve available data for analysis. They can use the advanced search options (6) to find certain traits, subjects, or data. Using menu option ‘ file|down- load ’ (7) they can download visible/selected (8) data as tab-delimited files to analyze them in third party software. Bioinformaticians can ‘ plug-in ’ a custom-built screen (see ‘ customization ’ section) that allows processing of selected data inside the GUI, for example, visua- lizing a correlation matrix as a graph (9) without the additional steps of downloading data and uploading it into another tool. Biologists can create link-outs to related information, for example, to probes in GeneNet- work.org (not shown). Table 5 summarizes use cases of the graphical user interface. De facto standard analysis tools are emerging, for example, tools for transcript data [20,21,24] or metabolite abundance data [22] to mention just a few. These tools are typically implemented using the open source software for statistical analysis and graphics named R [19]. Bioinformaticians can connect their particular R or Java programs to the XGAP database using an API with similar functionality to the GUI, that is, using simple commands like ‘ find ’ , ‘ add ’ and ‘ update ’ (R/API, Java/ API). Scripts in other programming languages and workflow tools like Taverna [53] can use \n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "unwanted_symbols = \"()[],‘’\"\n",
        "data = str.maketrans('','',unwanted_symbols)\n",
        "with open(\"/ganesh.txt\",'r') as file:\n",
        "  for line in file:\n",
        "    cleaned_data = line.translate(data)\n",
        "    print(cleaned_data)"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "AE-5gcIkvevb",
        "outputId": "7186ca4c-bc11-4b0c-8a36-1a0a79d70e77"
      },
      "execution_count": 24,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "For example a normalization Protocol may involve a  robust multiarray average RMA normalization  Action that uses Bioconductor  affy  Software 43 with certain Para- meterValues . Data types Description  BibliographicReferences  DatabaseEntry  URI  and FileAttachment enable researchers to freely add additional annotations to certain data types - DimensionElement  Investigation  Protocol  ProtocolApplication  and Data . For example researchers can annotate a Gene with one or more DatabaseEntries  referring to unique database accession numbers for automated data integration. A unique feature of XGAP is the uniform treatment of the various trait and subject annotations. The drawback of allowing users to freely add additional annotations such as described above is that users and tools using metabolite and gene traits for example would have to inspect each Trait instance to see whether it is actually a metabolite or gene and how it is annotated. That is why we instead use the object-oriented method of  inheritance  to explicitly add essential properties to Trait and Subject variants to make sure that they are described in a uniform way. For example Metabolite extends Trait  which explicitly adds properties ID Name and Type inherited from DimensionElement  to metabolite specific properties Mass Formula and Structure. See Jones et al. 38 for the complete FuGE specifi- cations and Jones and Paton 44 for a discussion on the benefits and drawbacks of alternative mechanisms for supporting extension in object models. Table 4 illustrates the usage of these annotation data types. Another feature of XGAP is the uniform treatment of all data on these subjects and traits. To understand basic data in XGAP newcomers just have to learn that all data are stored as Data matrices with each DataElement describing an observation on Subjects and/or Traits rows × columns. Unlike the proven matrix structures used in MAGE-TAB tabular format for microarray gene expression experiments 45 in XGAP these data can be on any Trait and/or Subject combination that is we did not create many variants of DataElement to accommodate each combination of Trait and Subject such as MAGE-TAB  s ExpressionDataElement Probe × Sample MassSpecDataElement MassPeak × Sample eQtlMappingDataElement Marker × Probe and so on. Instead we store all these data using the generic type DataElement and limit extension to Trait and Subject only. This avoids the combinatorial explo- sion of DataElement extensions so researchers can provide basic data as common data matrices of DataElements  and can still add particular annotations flexibly to the matrix row and columns to allow for new biotechnologies as demonstrated in the various Trait extensions in Figure 1. Keeping this simple and uniform data structure greatly enhances data and software reusability and hence productivity in line with the findings by Brazma et al. 29 and Rayner et al. 45 that the simple tabular structures underlying biological data should be exploited instead of making it overly complicated. After structural homogenization such as provided by FuGE and XGAP semantic queries are the remaining major barrier for integration of experimental metadata. This requires ontologies that describe the properties of the materials and also descriptions of experimental pro- cesses data and instruments. The former are provided by species-specific ontologies that are available from various sources. The Ontology for BioMedical investigation 46 may provide a solution for the experimental descriptors and is being used in this context by for example the Immune Epitope Database 47. To enable researchers to use these well understood descriptors XGAP inherits from FuGE the mechanism of  annotations   a special field to link any data object to one or more ontology terms. For example researchers can annotate a Gene with one or more OntologyTerms if required referring to standard ontology terms from OBO 39 or ontology terms defined locally. To enable data exchange using the XGAP model we produced a simple text-file format XGAP-TAB based on the experience that for data formats to be used data files should be easily created using simple Excel and text editor tools and closely resemble existing practices. This format is automatically derived from the model by requiring that all annotations on Investigations  Protocols  Traits  Subjects  and extensions thereof are described as delimited text files one file per data type with columns matching the properties described in the object model and each row describing one data instance. Optionally sets of DataElements can also be formatted as separate text matrices with row and column names matching these in the Trait and Subject annotation files and with each matrix value matching one DataElement . The dimensions of each data matrix are then listed by a row in the annotations on Data . Figure 2 shows one investigation in the XGAP tabular data format with one delimited text file per data type that is there are files named  probe.txt  and  individual. txt   with each row describing a microarray probe or individual respectively - and one text matrix file per set of DataElements - that is there are files named  data/ expressions.txt  and  data/genotypes.txt  . The properties of each data matrix is then described in  data.txt  ; that is for the  data/expressions.txt  there is a row in  data. txt  that says that its columns refer to  individual.txt   that its rows refer to  probe.txt  and that its values are  decimal  . Raw data sets and data sets in other formats can be retained in a directory labeled  original  . After proving its value in several proprietary projects a growing array of public data sets are now available at 48 demonstrating the use of XGAP-TAB 81113144950. A pilot software infrastructure is available at 51 to help genotype-to-phenotype researchers to adopt XGAP as a backbone for their data and tool integration. We chose to use the MOLGENIS toolkit biosoftware generator for MOLecular GENetics Information Systems; see Materials and methods to auto-generate from the XGAP model: 1 an SQL Structured Query Language for relational databases file with all necessary statements for setting up your own customized variant of the XGAP database; 2 application programming interfaces APIs in R Java and Web Services that allow bioinformaticians to plug-in their R processing scripts Taverna workflows 255253 and other tools; 3 a bespoke web-based graphical user interface GUI by which researchers can submit and retrieve data and run plugged-in tools; and 4 import/export wizards to un load and validate data sets exchanged in XGAP-TAB format. The auto-generation process can be repeated to quickly customize XGAP from an extended model for example to accommodate a particular new type of measurement technology or experimental design. Figure 3 shows the GUI to upload manage find and download genotype and phenotype data to the database. The GUI is generated with a uniform  look-and-feel   thereby lowering the barrier for novice users. Investigations can be described with all subjects traits data and protocol applications involved 1. The numbers refer to steps in the figure. Data can be entered using either the edit boxes or using menu-option  file|upload  2. This option enables upload of whole lists of traits and subjects from a simple tab-delimited format 3 which can easily be produced with Excel or R; MOLGENIS automatically generates online documentation describing the expected format 4. Subsequently the protocol applications involved can be added with the resulting raw data for example genetic fingerprints expression profiles and processed data for example normalized profiles QTL profiles metabolic networks. These data can be uploaded again using the common tab-delimited format or custom parsers 5 that bioinformaticians can  plug-in  for specific file formats for example Affymetrix CEL files. The software behind the GUI checks the relationships between subjects traits and data elements so no  orphaned  data are loaded into the database - for example genetic fingerprint data cannot be added before all information is uploaded on the markers and subjects involved. Standard paths through the data upload process are employed to ensure that only complete and valid data are uploaded and to provide a consistent user experience. Biologists can use the graphical user interface to navi- gate and retrieve available data for analysis. They can use the advanced search options 6 to find certain traits subjects or data. Using menu option  file|down- load  7 they can download visible/selected 8 data as tab-delimited files to analyze them in third party software. Bioinformaticians can  plug-in  a custom-built screen see  customization  section that allows processing of selected data inside the GUI for example visua- lizing a correlation matrix as a graph 9 without the additional steps of downloading data and uploading it into another tool. Biologists can create link-outs to related information for example to probes in GeneNet- work.org not shown. Table 5 summarizes use cases of the graphical user interface. De facto standard analysis tools are emerging for example tools for transcript data 202124 or metabolite abundance data 22 to mention just a few. These tools are typically implemented using the open source software for statistical analysis and graphics named R 19. Bioinformaticians can connect their particular R or Java programs to the XGAP database using an API with similar functionality to the GUI that is using simple commands like  find    add  and  update  R/API Java/ API. Scripts in other programming languages and workflow tools like Taverna 53 can use \n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "?str.maketrans"
      ],
      "metadata": {
        "id": "Blem0IS80LEB"
      },
      "execution_count": 25,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [],
      "metadata": {
        "id": "ZdcJCGlq5H9v"
      },
      "execution_count": null,
      "outputs": []
    }
  ]
}